### PR TITLE
verify that NSTemplateTier are created/updated at startup

### PR DESCRIPTION
**Pull request by xcoulon**
add test to verify that the NSTemplateTier resources
are created and updated (for that, a 'basic' NSTemplateTier
resource is created before the host-operator controller
is deployed)

also: small refactoring in the makefile (adding comments
and removing an empty goal)

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

File: make/test.mk [hunk 1]
```diff
@@ -178,6 +178,9 @@ ifneq ($(IS_OS_3),)
 	cat ${HOST_REPO_PATH}/deploy/cluster_role_binding.yaml | sed s/\REPLACE_NAMESPACE/$(HOST_NS)/ | oc apply -f -
 	oc apply -f ${HOST_REPO_PATH}/deploy/crds
 endif
+	# also, add a single `NSTemplateTier` resource before the host-operator controller is deployed. This resource will be updated
+	# as the controller starts (which is a use-case for CRT-231)
```
Comment (alexeykazakov) [inline make/test.mk]: Nice trick :)

File: test/e2e/nstemplatetier_test.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	"github.com/codeready-toolchain/toolchain-e2e/wait"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	client "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func TestCreateNSTemplateTierAtStartup(t *testing.T) {
+	// given:
+	// given
```
Comment (alexeykazakov) [inline test/e2e/nstemplatetier_test.go]: two `given`s

File: test/e2e/nstemplatetier_test.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	// then
+	require.NoError(t, err)
+	assert.Len(t, tierList.Items, 2)
+	for _, tier := range tierList.Items {
+		switch tier.Name {
+		case "advanced":
+			// the `advanced` tier should be in generation 1 (just created)
+			assert.Equal(t, int64(1), tier.ObjectMeta.GetGeneration())
+		case "basic":
+			// the `basic` tier should be in generation 1 (just updated - was created by makefile)
```
Comment (alexeykazakov) [inline test/e2e/nstemplatetier_test.go]:
```suggestion
			// the `basic` tier should be in generation 2 (just updated - was created by makefile)
```

File: test/e2e/nstemplatetier_test.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+func TestCreateNSTemplateTierAtStartup(t *testing.T) {
+	// given:
+	// given
+	tierList := &toolchainv1alpha1.NSTemplateTierList{}
+	ctx, awaitility := doubles.InitializeOperators(t, tierList)
+	defer ctx.Cleanup()
+	hostNs := os.Getenv(wait.HostNsVar)
+
+	// when
+	err := awaitility.Client.List(context.TODO(), client.InNamespace(hostNs), tierList)
```
Comment (alexeykazakov) [inline test/e2e/nstemplatetier_test.go]: You can't list the objects right away. They are created/updated asynchronously. You have to wait for them. Take a look at https://github.com/codeready-toolchain/toolchain-e2e/blob/84651717e851f21de951bc30590f3dee9be1d7ba/test/e2e/e2e_test.go#L205 for example.

File: make/test.mk (hunk 1 above)
Comment (xcoulon) [reply make/test.mk]: thanks ;)

File: test/e2e/nstemplatetier_test.go (hunk 2 above)
Comment (xcoulon) [reply test/e2e/nstemplatetier_test.go]: is it too much? ;)
removed duplicate comment in a788e17

File: test/e2e/nstemplatetier_test.go (hunk 4 above)
Comment (xcoulon) [reply test/e2e/nstemplatetier_test.go]: ah ok, let me do that

Comment (xcoulon) [reply test/e2e/nstemplatetier_test.go]: test refactored in b605ffe

File: make/test.mk [hunk 5]
```diff
@@ -14,27 +14,27 @@ IS_CRC := $(shell oc config view --minify -o jsonpath='{.clusters[0].cluster.ser
 IS_KUBE_ADMIN := $(shell oc whoami | grep "kube:admin")
 IS_OS_3 := $(shell curl -k -XGET -H "Authorization: Bearer $(shell oc whoami -t 2>/dev/null)" $(shell oc config view --minify -o jsonpath='{.clusters[0].cluster.server}')/version/openshift 2>/dev/null | grep paths)
 
-.PHONY: test-e2e-keep-namespaces
-test-e2e-keep-namespaces: login-as-admin deploy-member deploy-host setup-kubefed e2e-run
-
```
Comment (MatousJobanek) [inline make/test.mk]: may I know why did you remove this target? I sometimes use it.

File: test/e2e/nstemplatetier_test.go [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+func TestCreateOrUpdateNSTemplateTierAtStartup(t *testing.T) {
+	// given
+	tierList := &toolchainv1alpha1.NSTemplateTierList{}
+	ctx, awaitility := doubles.InitializeOperators(t, tierList)
+	defer ctx.Cleanup()
+	hostAwait := wait.NewHostAwaitility(awaitility)
+
+	// check the "advanced" NSTemplateTier, expecting generation=1 (just created)
+	err := hostAwait.WaitForNSTemplateTier("advanced", wait.NSTemplateTierHavingGeneration(1))
```
Comment (MatousJobanek) [inline test/e2e/nstemplatetier_test.go]: I'm not sure if checking generation is really a good approach - that is something out of our control, isn't it? Wouldn't be better to only check the existence of the tier in this case?

File: test/e2e/nstemplatetier_test.go [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+	require.NoError(t, err)
+
+	// check the "basic" NSTemplateTier, expecting generation=2 (just updated)
+	err = hostAwait.WaitForNSTemplateTier("basic", wait.NSTemplateTierHavingGeneration(2))
```
Comment (MatousJobanek) [inline test/e2e/nstemplatetier_test.go]: And in this case, verify that the revisions of the namespaces don't equal to `000000a`

File: make/test.mk (hunk 5 above)
Comment (xcoulon) [reply make/test.mk]: my bad, I can restore it

File: test/e2e/nstemplatetier_test.go (hunk 6 above)
Comment (xcoulon) [reply test/e2e/nstemplatetier_test.go]: well, there's already an older tier, so I would need to make sure that it was updated first... or should I check the content instead? Eg, make sure that the `namespace[].revision` matches a given value? But since we start with `generation=1`, can't we just assume that `generation=2` is for the updated object?

File: test/e2e/nstemplatetier_test.go (hunk 7 above)
Comment (xcoulon) [reply test/e2e/nstemplatetier_test.go]: ok, I can do that, indeed

Comment (xcoulon) [reply test/e2e/nstemplatetier_test.go]: done in 7a4b027

File: make/test.mk (hunk 5 above)
Comment (xcoulon) [reply make/test.mk]: done in 325bc5a

File: test/e2e/nstemplatetier_test.go (hunk 6 above)
Comment (xcoulon) [reply test/e2e/nstemplatetier_test.go]: done in 7a4b027

File: test/e2e/nstemplatetier_test.go [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+	defer ctx.Cleanup()
+	hostAwait := wait.NewHostAwaitility(awaitility)
+
+	// check the "advanced" NSTemplateTier exists (just created)
+	err := hostAwait.WaitForNSTemplateTier("advanced")
+	require.NoError(t, err)
+
+	// check the "basic" NSTemplateTier exists, and all its Namespace revisions are different from `000000a`,
+	// which is the value specified in the initial manifest
+	err = hostAwait.WaitForNSTemplateTier("basic", wait.NSTemplateTierSpecHaving(wait.Not(wait.NamespaceRevisions("000000a"))))
```
Comment (MatousJobanek) [inline test/e2e/nstemplatetier_test.go]: I like this one :+1:

Comment (xcoulon) [reply test/e2e/nstemplatetier_test.go]: thanks ;)

Comment (MatousJobanek) [reply test/e2e/nstemplatetier_test.go]: I had the same thinking that the dot import would improve the readability but I didn't want to bother you with additional changes inside of this PR. 
Anyway, +1 from my side for using dot imports

Comment (xcoulon) [reply test/e2e/nstemplatetier_test.go]: ok, let me do it ;)

Comment (xcoulon) [reply test/e2e/nstemplatetier_test.go]: done in f36a7bc
